File: packages/scheduling-engine/src/analysis/coverage-audit.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Coverage Audit
//  Per-operation demand vs. production verification
//  Extracted from NikufraEngine.tsx auditCoverage()
// ═══════════════════════════════════════════════════════════

import type { Block } from '../types/blocks.js';
import type { EOp, ETool } from '../types/engine.js';
import type { TwinGroup } from '../types/twin.js';
import { getBlockProductionForOp } from '../utils/block-production.js';

// ── Types ────────────────────────────────────────────────────

export interface CoverageAuditRow {
  opId: string;
  sku: string;
  nm: string;
  machineId: string;
  toolId: string;
  totalDemand: number;
  produced: number;
  coveragePct: number;
  gap: number;
  reason: 'ok' | 'overflow' | 'blocked' | 'partial' | 'rate_zero' | 'no_demand';
  hasAlt: boolean;
  altM: string | null;
  /** Whether production comes from twin co-production */
  isTwinProduction?: boolean;
  /** Twin partner op ID (if co-produced) */
  twinPartnerOpId?: string;
  /** Excess production from co-production going to stock */
  twinExcessToStock?: number;
}

export interface CoverageAuditResult {
  rows: CoverageAuditRow[];
  totalDemand: number;
  totalProduced: number;
  globalCoveragePct: number;
  fullyCovered: number;
  partiallyCovered: number;
  zeroCovered: number;
  isComplete: boolean;
}

// ── Main Computation ────────────────────────────────────────

/**
 * Audit demand coverage for each operation.
 *
 * For each operation, computes total demand (daily + backlog),
 * produced quantity from 'ok' blocks, and categorizes the result.
 */
export function auditCoverage(
  blocks: Block[],
  ops: EOp[],
  TM: Record<string, ETool>,
  twinGroups?: TwinGroup[],
): CoverageAuditResult {
  const rows: CoverageAuditRow[] = [];
  let totalDemand = 0;
  let totalProduced = 0;
  let fullyCovered = 0;
  let partiallyCovered = 0;
  let zeroCovered = 0;

  for (const op of ops) {
    const demand = op.d.reduce((s, v) => s + Math.max(v, 0), 0) + Math.max(op.atr, 0);
    const tool = TM[op.t];

    if (demand <= 0) {
      rows.push({
        opId: op.id,
        sku: op.sku,
        nm: op.nm,
        machineId: op.m,
        toolId: op.t,
        totalDemand: 0,
        produced: 0,
        coveragePct: 100,
        gap: 0,
        reason: 'no_demand',
        hasAlt: !!(tool?.alt && tool.alt !== '-'),
        altM: tool?.alt && tool.alt !== '-' ? tool.alt : null,
      });
      fullyCovered++;
      continue;
    }

    totalDemand += demand;

    // Check if rate=0 (scheduler skips these)
    if (!tool || tool.pH <= 0) {
      rows.push({
        opId: op.id,
        sku: op.sku,
        nm: op.nm,
        machineId: op.m,
        toolId: op.t,
        totalDemand: demand,
        produced: 0,
        coveragePct: 0,
        gap: demand,
        reason: 'rate_zero',
        hasAlt: !!(tool?.alt && tool.alt !== '-'),
        altM: tool?.alt && tool.alt !== '-' ? tool.alt : null,
      });
      zeroCovered++;
      continue;
    }

    const opBlocks = blocks.filter((b) => b.opId === op.id);
    // Twin-aware production: uses outputs[] for co-production blocks
    const produced = getBlockProductionForOp(blocks, op.id);
    const hasBlocked = opBlocks.some((b) => b.type === 'blocked');
    const hasOverflow = opBlocks.some((b) => b.type === 'overflow');

    // Twin co-production metadata
    const twinBlks = blocks.filter(
      (b) => b.type === 'ok' && b.isTwinProduction && b.outputs?.some((o) => o.opId === op.id),
    );
    const isTwin = twinBlks.length > 0;
    let twinPartnerOpId: string | undefined;
    let twinExcessToStock: number | undefined;
    if (isTwin && twinGroups) {
      const grp = twinGroups.find((g) => g.opId1 === op.id || g.opId2 === op.id);
      twinPartnerOpId = grp ? (grp.opId1 === op.id ? grp.opId2 : grp.opId1) : undefined;
      if (produced > demand) twinExcessToStock = produced - demand;
    }

    totalProduced += produced;
    const pct = Math.min(100, demand > 0 ? (produced / demand) * 100 : 100);
    const gap = Math.max(0, demand - produced);

    let reason: CoverageAuditRow['reason'] = 'ok';
    if (pct < 100) {
      reason = hasBlocked ? 'blocked' : hasOverflow ? 'overflow' : 'partial';
    }

    rows.push({
      opId: op.id,
      sku: op.sku,
      nm: op.nm,
      machineId: op.m,
      toolId: op.t,
      totalDemand: demand,
      produced,
      coveragePct: Math.round(pct),
      gap,
      reason,
      hasAlt: !!(tool?.alt && tool.alt !== '-'),
      altM: tool?.alt && tool.alt !== '-' ? tool.alt : null,
      ...(isTwin ? { isTwinProduction: true, twinPartnerOpId, twinExcessToStock } : {}),
    });

    if (produced >= demand) fullyCovered++;
    else if (produced > 0) partiallyCovered++;
    else zeroCovered++;
  }

  const globalPct = totalDemand > 0 ? Math.min(100, (totalProduced / totalDemand) * 100) : 100;

  return {
    rows,
    totalDemand,
    totalProduced,
    globalCoveragePct: Math.round(globalPct * 10) / 10,
    fullyCovered,
    partiallyCovered,
    zeroCovered,
    isComplete: totalProduced >= totalDemand,
  };
}

File: apps/backend/src/domain/scheduling/analysis/coverage_audit.py
"""Coverage audit — port of analysis/coverage-audit.ts.

Per-operation demand coverage verification.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..types import Block, EOp, ETool, TwinGroup


@dataclass
class CoverageAuditRow:
    op_id: str
    sku: str
    nm: str = ""
    machine_id: str = ""
    tool_id: str = ""
    total_demand: int = 0
    produced: int = 0
    coverage_pct: float = 0
    gap: int = 0
    reason: str = "ok"  # ok, overflow, blocked, partial, rate_zero, no_demand
    has_alt: bool = False
    alt_m: str | None = None
    is_twin_production: bool = False
    twin_partner_op_id: str | None = None
    twin_excess_to_stock: int = 0


@dataclass
class CoverageAuditResult:
    rows: list[CoverageAuditRow] = field(default_factory=list)
    total_demand: int = 0
    total_produced: int = 0
    global_coverage_pct: float = 0
    fully_covered: int = 0
    partially_covered: int = 0
    zero_covered: int = 0
    is_complete: bool = True


def _get_block_production_for_op(blocks: list[Block], op_id: str) -> int:
    total = 0
    for b in blocks:
        if b.type != "ok" or b.qty <= 0:
            continue
        if b.op_id == op_id:
            total += b.qty
        if b.outputs:
            for out in b.outputs:
                if out.op_id == op_id:
                    total += out.qty
    return total


def audit_coverage(
    blocks: list[Block],
    ops: list[EOp],
    tool_map: dict[str, ETool],
    twin_groups: list[TwinGroup] | None = None,
) -> CoverageAuditResult:
    """Audit per-op demand coverage."""
    rows: list[CoverageAuditRow] = []
    total_demand = 0
    total_produced = 0
    fully_covered = 0
    partially_covered = 0
    zero_covered = 0

    # Build twin partner map
    twin_partner: dict[str, str] = {}
    if twin_groups:
        for tg in twin_groups:
            twin_partner[tg.op_id1] = tg.op_id2
            twin_partner[tg.op_id2] = tg.op_id1

    for op in ops:
        demand = sum(max(v, 0) for v in op.d) + max(op.atr, 0)

        if demand <= 0:
            rows.append(
                CoverageAuditRow(
                    op_id=op.id,
                    sku=op.sku,
                    nm=op.nm,
                    machine_id=op.m,
                    tool_id=op.t,
                    total_demand=0,
                    produced=0,
                    coverage_pct=100.0,
                    gap=0,
                    reason="no_demand",
                )
            )
            fully_covered += 1
            continue

        total_demand += demand

        tool = tool_map.get(op.t)
        if not tool or tool.pH <= 0:
            rows.append(
                CoverageAuditRow(
                    op_id=op.id,
                    sku=op.sku,
                    nm=op.nm,
                    machine_id=op.m,
                    tool_id=op.t,
                    total_demand=demand,
                    produced=0,
                    coverage_pct=0,
                    gap=demand,
                    reason="rate_zero",
                )
            )
            zero_covered += 1
            continue

        produced = _get_block_production_for_op(blocks, op.id)
        total_produced += produced

        pct = min(100.0, (produced / demand * 100.0) if demand > 0 else 100.0)
        gap = max(0, demand - produced)

        # Classify reason
        has_blocked = any(b.op_id == op.id and b.type == "blocked" for b in blocks)
        has_overflow = any(b.op_id == op.id and b.overflow for b in blocks)

        if pct >= 100:
            reason = "ok"
        elif has_blocked:
            reason = "blocked"
        elif has_overflow:
            reason = "overflow"
        else:
            reason = "partial"

        # Twin metadata
        is_twin = any(b.op_id == op.id and b.is_twin_production and b.type == "ok" for b in blocks)
        tp_op_id = twin_partner.get(op.id)
        excess = max(0, produced - demand) if is_twin else 0

        has_alt = tool.alt != "-" if tool.alt else False
        alt_m = tool.alt if has_alt else None

        rows.append(
            CoverageAuditRow(
                op_id=op.id,
                sku=op.sku,
                nm=op.nm,
                machine_id=op.m,
                tool_id=op.t,
                total_demand=demand,
                produced=produced,
                coverage_pct=round(pct, 1),
                gap=gap,
                reason=reason,
                has_alt=has_alt,
                alt_m=alt_m,
                is_twin_production=is_twin,
                twin_partner_op_id=tp_op_id,
                twin_excess_to_stock=excess,
            )
        )

        if produced >= demand:
            fully_covered += 1
        elif produced > 0:
            partially_covered += 1
        else:
            zero_covered += 1

    global_pct = min(100.0, (total_produced / total_demand * 100.0) if total_demand > 0 else 100.0)

    return CoverageAuditResult(
        rows=rows,
        total_demand=total_demand,
        total_produced=total_produced,
        global_coverage_pct=round(global_pct, 1),
        fully_covered=fully_covered,
        partially_covered=partially_covered,
        zero_covered=zero_covered,
        is_complete=total_produced >= total_demand,
    )
